llvm/lib/Transforms/InstCombine/InstCombineCompares.cpp

```diff
@@ -447,11 +449,11 @@ static bool canRewriteGEPAsOffset(Type *ElemTy, Value *Start, Value *Base,
         return false;
 
       if (auto *GEP = dyn_cast<GEPOperator>(V)) {
-        // We're limiting the GEP to having one index. This will preserve
-        // the original pointer type. We could handle more cases in the
-        // future.
-        if (GEP->getNumIndices() != 1 || !GEP->isInBounds() ||
-            GEP->getSourceElementType() != ElemTy)
+        // Only allow GEPs with at most one variable offset.
+        APInt Offset(IndexSize, 0);
+        MapVector<Value *, APInt> VarOffsets;
+        if (!GEP->collectOffset(DL, IndexSize, VarOffsets, Offset) ||
+            VarOffsets.size() > 1)
```

Comment:
Should this be != 1? Or is zero-index GEP truly impossible?

Reply:
VarOffsets.empty() means that the GEP has a constant offset, not that it doesn't have indices.

(As a side-note, zero-index GEPs are indeed allowed, but they are always no-ops and optimized away.)

Reply:
Why do you not sum constant offsets? Also probably best to switch order as the `size` check will be cheaper than `collectOffset`.

Reply:
I've changed this code to just explicitly count the non-constant indices. There was really no need to use collectOffset() for this purpose here.